File: go/internal/sol/p0050/p0050.go
package p0050

import (
	"strconv"

	"pe-solver/internal/mylib"
)

type csumGenerator struct {
	p    int
	csum int
}

func (cs *csumGenerator) next() int {
	cs.p = mylib.NextPrime(cs.p)
	cs.csum += cs.p

	return cs.csum
}

func initCsumLst(csumGen *csumGenerator, limit int) []int {
	lst := []int{0}
	for lst[len(lst)-1] < limit {
		lst = append(lst, csumGen.next())
	}

	return lst
}

func compute(limit int) string {
	if limit < 3 {
		panic("limit must be larger than 2")
	}

	csumGen := new(csumGenerator)
	lst := initCsumLst(csumGen, limit)

	k := len(lst) - 2
	left := 0
	for {
		diff := lst[left+k] - lst[left]
		if diff >= limit {
			left = 0
			k -= 1
		} else if !mylib.IsPrime(diff) {
			left += 1
			if left+k >= len(lst) {
				lst = append(lst, csumGen.next())
			}
		} else {
			return strconv.FormatInt(int64(diff), 10)
		}
	}

	panic("unreachable")
}

func Solve() string {
	return compute(1_000_000)
}

File: haskell/src/Sol/P0050.hs
module Sol.P0050 (compute, solve) where

import Mylib.Prime (isPrime, primeNumbers)

-- 0, 2, 5, 10, 17, 28, 41, 58, 77, 100, 129, ...
cumsumPrimes :: [Int]
cumsumPrimes = scanl (+) 0 primeNumbers

compute :: Int -> String
compute limit =
    show $ go 0 max_offset
  where
    max_offset = pred . length $ takeWhile (< limit) cumsumPrimes

    go :: Int -> Int -> Int
    go left k
        | diff >= limit =
            go 0 (pred k)
        | isPrime diff =
            diff
        | otherwise =
            go (succ left) k
      where
        diff = cumsumPrimes !! (left + k) - cumsumPrimes !! left

solve :: String
solve = compute 1_000_000
